File: LeetCode/138-copy-list-with-random-pointer/ex_138.py
# Definition for singly-linked list with a random pointer.
# class RandomListNode(object):
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None

class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: RandomListNode
        :rtype: RandomListNode
        """
        if not head:    return
        p = head
        while p:
            dup = RandomListNode(p.label)
            dup.next = p.next
            p.next = dup
            p = dup.next
        p = head
        while p:
            dup = p.next
            if p.random:
                dup.random = p.random.next
            p = dup.next
        p = head
        dup_head = head.next
        while p:
            dup = p.next
            p.next = dup.next
            if dup.next:
                dup.next = dup.next.next
            p = p.next
        return dup_head

File: LeetCode/138-copy-list-with-random-pointer/ex_138.cpp
/**
 * Definition for singly-linked list with a random pointer.
 * struct RandomListNode {
 *     int label;
 *     RandomListNode *next, *random;
 *     RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
 * };
 */
class Solution {
public:
    RandomListNode *copyRandomList(RandomListNode *head) {
        map<RandomListNode*, RandomListNode*> nodes;
        auto curr = head;
        while (curr) {
            auto copy = new RandomListNode(curr->label);
            nodes[curr] = copy;
            curr = curr->next;
        }
        curr = head;
        while (curr) {
            nodes[curr]->next = nodes[curr->next];
            nodes[curr]->random = nodes[curr->random];
            curr = curr->next;
        }
        return nodes[head];
    }
};